geoh5py/ui_json/enforcers.py

```diff
@@ -0,0 +1,350 @@
+    """
+    Enforces type validation.
+
+    :param validations: Valid type(s) for parameter value.
+    :raises TypeValidationError: If value is not a valid type.
+    """
+
+    enforcer_type: str = "type"
+
+    def __init__(self, validations: type | list[type]):
+        super().__init__(validations)
+
+    def enforce(self, name: str, value: Any):
+        """Administers rule to enforce type validation."""
+        if not self.rule(value):
+            raise TypeValidationError(
+                name, type(value).__name__, self._stringify(self.validations)
+            )
+
+    def rule(self, value) -> bool:
+        """True if value is one of the valid types."""
+        return any(isinstance(value, k) for k in self.validations)
+
+    @property
+    def validations(self) -> list[type]:
+        if not isinstance(self._validations, list):
+            self._validations = [self._validations]
+
+        return self._validations
+
+    def _stringify(self, target: type | list[type]) -> str | list[str]:
+        """Converts type(s) to string for error message."""
+
+        if isinstance(target, list):
+            return [k.__name__ for k in target]
+
+        if isinstance(target, type):
+            return target.__name__
+
+        raise TypeError(f"Cannot stringify type {type(target)}")
+
+
+class ValueEnforcer(Enforcer):
+    """
+    Enforces value validation.
+
+    :param validations: Valid parameter value(s).
+    :raises ValueValidationError: If value is not a valid value
+        choice.
+    """
+
+    enforcer_type = "value"
+
+    def __init__(self, validations: list[Any]):
+        super().__init__(validations)
+
+    def enforce(self, name: str, value: Any):
+        """Administers rule to enforce value validation."""
+        if not self.rule(value):
+            raise ValueValidationError(name, value, self.validations)
+
+    def rule(self, value: Any) -> bool:
+        """True if value is a valid choice."""
+        return value in self.validations
+
+
+class UUIDEnforcer(Enforcer):
+    """
+    Enforces uuid validation.
+
+    :param validations: None is considered a valid uuid if
+        validations is 'optional'.
+    :raises UUIDValidationError: If value is not a valid uuid string.
+    """
+
+    enforcer_type = "uuid"
+
+    def __init__(self, validations: str | None = None):
+        super().__init__(validations)
+
+    def enforce(self, name: str, value: Any):
+        """Administers rule to check if valid uuid."""
+        if not self.rule(value):
+            raise UUIDValidationError(
+                name,
+                value,
+            )
+
+    def rule(self, value: Any) -> bool:
+        """True if value is a valid uuid string."""
+
+        if self.validations == "optional":
+            return True
+
+        if not isinstance(value, str):
+            return False
+
+        try:
+            uuid.UUID(value)
+            is_uuid = True
+        except ValueError:
+            is_uuid = False

```

Comment:
I think we could re-use `shared.utils.is_uuid`

Reply:
Yes - thank you!